File: packages/java/src/main/java/io/opensms/Templates.java
package io.opensms;

import io.opensms.models.Page;
import io.opensms.models.Template;

/** Manage message templates. Accessed as {@code client.templates()}. */
public final class Templates {

    private final ApiTransport t;

    Templates(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of templates. */
    public Page<Template> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<Template> list(ListParams params) {
        return t.get("/v1/templates", params.toQuery(), ApiTransport.pageOf(Template.class));
    }

    /**
     * @param params {@code name} and {@code body} required.
     * @return the created template.
     */
    public Template create(TemplateParams params) {
        return create(params, null);
    }

    /**
     * @param params  {@code name} and {@code body} required.
     * @param options per-call options.
     * @return the created template.
     */
    public Template create(TemplateParams params, RequestOptions options) {
        Wire.require(params.name(), "name");
        Wire.require(params.body(), "body");
        return t.json("POST", "/v1/templates", params.toWire(), ApiTransport.idempotencyKey(options),
                ApiTransport.type(Template.class));
    }

    /**
     * @param id template id.
     * @return the template.
     */
    public Template get(String id) {
        return t.get("/v1/templates/" + Wire.id(id, "id"), null, ApiTransport.type(Template.class));
    }

    /**
     * @param id     template id.
     * @param params fields to change.
     * @return the updated template.
     */
    public Template update(String id, TemplateParams params) {
        return t.json("PATCH", "/v1/templates/" + Wire.id(id, "id"), params.toWire(), null,
                ApiTransport.type(Template.class));
    }

    /** @param id template id. */
    public void delete(String id) {
        t.json("DELETE", "/v1/templates/" + Wire.id(id, "id"), null, null, ApiTransport.VOID);
    }
}

File: packages/php/src/Resources/Templates.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `templates` resource: message templates with `{{name}}` placeholders. Fields: `name`, `body`, `traffic_type`.
 * Accessed as `$client->templates`.
 */
final class Templates
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Query: `limit` (1..200), `cursor`.
     *
     * @param array<string, mixed> $params
     */
    public function list(array $params = []): Page
    {
        $query = Models::map($params, Models::PAGE_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/templates', $query));
    }

    /**
     * @param array<string, mixed> $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function create(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::TEMPLATE, ['name', 'body']);

        return Models::object($this->transport->request('POST', '/v1/templates', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/templates/' . Models::pathParam($id)));
    }

    /**
     * Partial update (PATCH): only the given fields change.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function update(string $id, array $params): array
    {
        $body = Models::map($params, Models::TEMPLATE);

        return Models::object($this->transport->request('PATCH', '/v1/templates/' . Models::pathParam($id), [], $body));
    }

    public function delete(string $id): void
    {
        $this->transport->request('DELETE', '/v1/templates/' . Models::pathParam($id));
    }
}
